### ISSUE TITLE
change the way lag variables are detected in predict

### ISSUE THREAD

**Issue by rsbivand** [r-spatial/spatialreg#38]
Handling of formula Durbin objects does not work, get to line 233 in R/predict.sarlm.R.

**Comment by dpr01** [comment]
Hi, I have this issue when using `predict` on a Durbin model with only some lagged variables. Assuming the problem isn't easy to fix, would there be any other way to generate forecasts with a model like this? Thanks.

**Comment by rsbivand** [comment]
Could you provide a reproducible example that matches your case?

**Comment by dpr01** [comment]
So after looking at it more carefully, I believe my error was actually being caused by a mismatch of structure between my original dataset and the predicting dataset. The message I had was:

"Error in predict.Sarlm(mod, dfp, listw = weights) : 
  unknown mismatch. please report this bug
In addition: Warning message:
In colnames(Xs.not.lagged) != colnames(Xo) :
  longer object length is not a multiple of shorter object length"

I assumed the first error (which in the code you link to this issue) was the main problem but fixing the second warning appears to have made everything work. An example of what I'm doing:

```r
library(tidyverse)
library(sf)
library(spdep)
library(spatialreg)

# Dataset
df <- st_read(system.file("shapes/boston_tracts.shp", package = "spData"))

# Spatial weights
centroids <- st_centroid(df)
neigh <- dnearneigh(centroids, d1 = 0, d2 = 30)
weights <- nb2listw(neigh, glist = lapply(nbdists(neigh, centroids), \(x) 1/(x^2)))

# Model
mod <- lagsarlm(log(MEDV) ~ CRIM + ZN + INDUS + CHAS + I(NOX^2) + I(RM^2) +
                  AGE + log(DIS) + log(RAD) + TAX + PTRATIO + B + log(LSTAT), 
                data = df,
                Durbin = ~ CRIM,
                listw = weights)
summary(mod)

# Predict
dfp <- df %>% 
  mutate(TAX = 0.5 * TAX)

predict(mod, dfp, listw = weights)
``` 
I think this works fine. But just to be clear, `predict` is aware that only `CRIM` is a spatially lagged variable in the model, as specified in `lagsarlm`, right? I think I may have misunderstood this issue.

**Comment by rsbivand** [comment]
Yes, `predict` only uses `lag.CRIM`, not other lagged X variables.
